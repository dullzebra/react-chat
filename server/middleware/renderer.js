import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { SheetsRegistry } from 'react-jss/lib/jss';
import JssProvider from 'react-jss/lib/JssProvider';
import { createGenerateClassName } from 'material-ui/styles';

import configureStore from '../../src/store';
import App from '../../src/components/App';

const sheetsRegistry = new SheetsRegistry();
const generateClassName = createGenerateClassName();

const path = require('path');
const fs = require('fs');

export default (req, res) => {
  // point to the html file created by CRA's build tool
  const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');
  const context = {};
  const store = configureStore();

  fs.readFile(filePath, 'utf8', (err, htmlData) => {
    if (err) {
      console.error('err', err);
      return res.status(404).end();
    }

    if (context.url) {
      res.redirect(context.url);
      return;
    }

    // render the app as a string
    /* eslint-disable-next-line */
    const html = ReactDOMServer.renderToString(
      <JssProvider registry={sheetsRegistry} generateClassName={generateClassName}>
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            <App />
          </StaticRouter>
        </Provider>
      </JssProvider>);

    const css = sheetsRegistry.toString();

    // inject the rendered app into our html and send it
    return res.end(htmlData.replace(
      '<div id="root"></div>',
      `<div id="root">${html}</div>
    <script>window.INITIAL_STATE = ${JSON.stringify(store.getState()).replace(
    /</g,
    '\\u003c',
  )}</script>
    <style id="jss-server-side">${css}</style>
    `,
    ));
  });
};
