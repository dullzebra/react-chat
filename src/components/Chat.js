import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';
import Typography from 'material-ui/Typography';
import * as style from '../constants/styles';
import ChatMessageList from './ChatMessageList';
import ChatInput from './ChatInput';

const styles = theme => ({
  content: {
    flexGrow: 1,
    display: 'flex',
    marginTop: style.appBarHeight,
    backgroundColor: theme.palette.background.default,
  },
  chatContainer: {
    flexGrow: 1,
    display: 'flex',
    flexDirection: 'column',
    overflow: 'auto',
  },
  messageList: {
    position: 'relative',
    flexBasis: '100%',
  },
  chatIntro: {
    flexBasis: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
});

class Chat extends React.Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    sendMessage: PropTypes.func.isRequired,
    joinChat: PropTypes.func.isRequired,
    messages: PropTypes.array.isRequired,
    activeChat: PropTypes.object,
    user: PropTypes.shape({
      _id: PropTypes.string,
    }).isRequired,
    isConnected: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    activeChat: null,
  };

  componentDidMount() {
    this.scrollDownHistory();
  }

  componentDidUpdate() {
    this.scrollDownHistory();
  }

  scrollDownHistory() {
    if (this.messagesContainer) {
      this.messagesContainer.scrollTop = this.messagesContainer.scrollHeight;
    }
  }

  handleSend = (content) => {
    this.props.sendMessage(content);
  };

  render() {
    const {
      classes, user, activeChat, joinChat, messages, isConnected,
    } = this.props;

    const msgPanel = messages.length ? (
      <ChatMessageList messages={messages} userId={user._id} />
    ) : (
      <div className={classes.chatIntro}>
        <Typography variant="subheading">В чате пока тихо... Будь первым!</Typography>
      </div>
    );

    const footer = user.isChatMember ? (
      <ChatInput disabled={!isConnected} sendMessage={this.handleSend} />
    ) : (
      <Button
        disabled={!isConnected}
        onClick={joinChat}
        variant="raised"
        color="primary"
        style={{ margin: 16 }}
      >
        Хочу к вам
      </Button>
    );

    const unactivePanel = (
      <div className={classes.chatIntro}>
        <Typography variant="subheading">Выбери чат из списка слева или создай свой</Typography>
      </div>
    );

    return (
      <main className={classes.content}>
        <div
          ref={(c) => {
            this.messagesContainer = c;
          }}
          className={classes.chatContainer}
        >
          {activeChat ? (
            <React.Fragment>
              {msgPanel}
              {footer}
            </React.Fragment>
          ) : (
            unactivePanel
          )}
        </div>
      </main>
    );
  }
}
export default withStyles(styles)(Chat);
