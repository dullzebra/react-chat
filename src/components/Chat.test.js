/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Chat from './Chat';

jest.mock('./ChatMessageList', () => () => 'ChatMessageList');
jest.mock('./ChatInput', () => () => 'ChatInput');

const mockData = {
  user: {
    _id: '12344',
  },
  activeChat: {},
  joinChat: jest.fn(),
  messages: [{}],
  sendMessage: jest.fn(),
  isConnected: true,
};

describe('<Chat />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Chat {...mockData} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Chat {...mockData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
