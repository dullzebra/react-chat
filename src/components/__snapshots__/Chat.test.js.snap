// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> renders correctly 1`] = `
<main
  className="Chat-content-29"
>
  <div
    className="Chat-chatContainer-30"
  >
    ChatMessageList
    <button
      className="MuiButtonBase-root-48 MuiButton-root-33 MuiButton-raised-38 MuiButton-raisedPrimary-40"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "margin": 16,
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label-34"
      >
        Хочу к вам
      </span>
      <span
        className="MuiTouchRipple-root-50"
      />
    </button>
  </div>
</main>
`;
